### ISSUE TITLE
Modals Panels losing focus and going behind main window

### ISSUE THREAD

**Issue by sdinesh86** [Adobe-CEP/CEP-Resources#117]
Hello,

I have created a simple toolbar like extension for AE.
I have been facing a very annoying issue with all Panels / Models ( Native and HTML Extensions ) losing focus and going behind the main window.
This issue only happens when I open my Panel.

Enviornment :
OS : Windows 7.
Application : Adobe After Effects CC2015 13.7.1
CEP : 6.1

Any help would be much appreciated.

Thanks,

**Comment by bbb999** [comment]
What specific steps lead to your panel being opened behind the main window?

**Comment by sdinesh86** [comment]
there are not specific steps this happens everytime i open any panel or modal in AE

**Comment by bbb999** [comment]
Bizarre. Do you see the same behavior using the AfterEffectsPanel sample? Does your panel display the same behavior, if you dock it into a workspace, then close/reopen AE?

**Comment by sdinesh86** [comment]
I don't see the same behavior using the AdobeKuler panel and everything works correctly if I close my panel.

I think there is something wrong I am doing in the extension which is causing this behavior.
Just one more detail.
My extension is like a pluggable system, the core of the extension is a browserified script with node modules, this script loads plugins from other folders ( using nodes require functionality ). Plugins can create views and display them using the core script.

This is the basic architecture of my extension, do you think this is causing the behavior.

**Comment by bbb999** [comment]
That sounds very complex. What workflow(s) does that enable, that wouldn't be available to a simple extension?

**Comment by sdinesh86** [comment]
this workflow enables creating plugable modules for extension, in this way i can i create paclages for extension.
Then in extension i can select which packages i would like to use and activate them. this workflow is inspired by atom editor.
All my packages are uploaded to local npm repo and downloaded from them into the extension folder, user will have the ability to browse, install and use packages he likes.

**Comment by bbb999** [comment]
That sounds ambitious! :) 

So, are the users downloading compiled plug-ins, or any compiled code modules which would need to be enumerated/known about, by your plug-in componentry, at app launch?

**Comment by sdinesh86** [comment]
> That sounds ambitious! :)

Yeah it is, and I don't know if I am doing it right and will be able to make it up to the level, but it doesnt hurt to try ;)

> So, are the users downloading compiled plug-ins, or any compiled code modules

Plugins are not compiled, they are just simple node modules ( folder with scripts and package.json )
currently, users are not downloading any plugins now since the functionality still doesn't exist, as its in initial phase. I am bundling a few modules with the installation, but later users will be able to download and manage the plugins.

> or any compiled code modules which would need to be enumerated/known about, by your plug-in componentry, at app launch?

Yes, modules are enumerated by the package management system and then are loaded and activated, at the launch of the application.

Also, there is one more concern I had, when I open my extension panel, it takes some time to show up.
I noticed this happens due to the loading of the packages, is there a way I can load the packages without blocking the panel and UI thread.

**Comment by bbb999** [comment]
Any node modules that the user downloaded during a given PPro session, would not show up until the _next_ PPro session. 

Panel opening time is dependent on file count, and Node.js is one of the best ways to increase file count. :) This is driven by signature checking of each file. Test: turn on PlayerDebugMode, and if the load time greatly improves, then the signature checking is the delay. In what PPro version are you testing? We made some significant panel load time improvements between 10.x and 11.0.

Downloading anything into the panel's directory will invalidate its signature, so PPro will no longer load it.

**Comment by sdinesh86** [comment]
My extension is only supposed to work on After Effects and Photoshop, so PPro is out of the question in this case.

I already have PlayerDebugMode turned on and it is the same case for every user.

A bit of loading is ok and manageable only if I can give feedback to user, but in my current case, i am not able to do it since panel doesn't show up until all the modules are loaded :(

* one more point during development I use live reload, which also reloads all the modules and practically restarts the whole extension, and shockingly this is instant ( 600ms ).
* Whereas closing and restarting starting an extension takes up around 2 - 3 seconds.
Note: ** when the modules are reloaded they are cleared from the cache and then required again.

I suspect that panel thread is blocked until the script execution is finished.
 Is there a workaround or solution for this.

**Comment by sdinesh86** [comment]
BUMP

**Comment by KupinFB** [comment]
Hi. I had a same issue. It can be because of using of standart HTML "title" atribute. If you have the "title" and if it was shown once, you will have alot of problems with Z-index of Panels and Modal Windows. 
I use not standart "tooltip libs" to show titles.
I will be glad if it helps you!
